File: interleaving.c
#include <stdio.h>
#include <unistd.h>    // for write
#include <pthread.h>   // for pthreads
#include <stdlib.h>    // for exit

// Print out an error message and exit.
static void fail( char const *message ) {
  fprintf( stderr, "%s\n", message );
  exit( 1 );
}

// Define the start routines for your three threads.  Each thread will
// use two calls to write() to print its two characters, then it will
// terminate.

void *threadFunc( void *arg) {
  
  if(write(1, *(char *)arg ,1) != 1)
    fail( "Can't write arg 1" );
  if(write(1, *(char *)arg + 1 ,1) != 1)
    fail( "Can't write arg 2" );
  return NULL;
}

int main( int argc, char *argv[] ) {
  // A bunch of times.
  for ( int i = 0; i < 100000; i++ ) {
    // Make three threads.

    pthread_t thread[ 3 ];
    char threadArg[ 2 ];
    threadArg[0] = 'a';
    threadArg[1] = 'b';
    if ( pthread_create( &(thread[ 0 ]), NULL, threadFunc, threadArg) )
      fail( "Can't make child thread" );
    threadArg[0] = 'a';
    threadArg[1] = 'b';
    if ( pthread_create( &(thread[ 0 ]), NULL, threadFunc, threadArg) )
      fail( "Can't make child thread" );
    threadArg[0] = 'a';
    threadArg[1] = 'b';
    if ( pthread_create( &(thread[ 0 ]), NULL, threadFunc, threadArg) )
      fail( "Can't make child thread" );
    
    // Join with the three threads.
    
    pthread_join( thread[ 0 ], NULL );
    pthread_join( thread[ 1 ], NULL );
    pthread_join( thread[ 2 ], NULL );
    
    // Use the write system call to print out a newline.  The string
    // we're passing to write is null terminated (since that's what
    // double quotesd strings are in C), but we're just using the
    // first byte (the newline).  Write doesn't care about null
    // terminated strings, it just writes out any sequence of bytes
    // you ask it to.
    write( STDOUT_FILENO, "\n", 1 );
  }

  return 0;
}

File: interleaving.java
public class Interleaving {
  /** Thread to print ab. */

static class MyRunnable implements Runnable {
  public void run() {
    System.out.print("ab");
  }
}

  /** Thread to print cd. */

static class MyRunnable1 implements Runnable {
  public void run() {
    System.out.print("cd");
  }
}


  /** Thread to print ef. */

static class MyRunnable2 implements Runnable {
  public void run() {
    System.out.print("ef");
  }
}

  public static void main( String[] args ) {
    // The three threads we make.
    Thread tab, tcd, tef;
  
    // A bunch of times.
    for ( int i = 0; i < 50000; i++ ) {
      // Make one of each type of threads.
      
      tab = new Thread( new MyRunnable());
      tcd = new Thread( new MyRunnable1());
      tef = new Thread( new MyRunnable2());
    
      // Start them all.

      tab.start();
      tcd.start();
      tef.start();

      // Join with our three threads.

    try {
      tab.join();
      tcd.join();
      tef.join();
    } catch ( InterruptedException e ) {
      System.out.println( "Can't join thread" );
    }

      // Print a newline to end the line they just printed.
      System.out.println();
    }
  }
}